Installation issue: fftw

### Steps to reproduce the issue

I cannot install `fftw +mpi` on macOS any more. The FFTW libraries are built without linking against the MPI libraries, leading to an error due to undefined symbols.

```console
$ spack --env-dir utils/spack/redshift spec -I fftw
Input spec
--------------------------------
 -   fftw

Concretized
--------------------------------
 -   fftw@3.3.10%apple-clang@13.0.0+mpi~openmp~pfft_patches precision=double,float arch=darwin-monterey-skylake
 -       ^openmpi@4.1.2%apple-clang@13.0.0~atomics~cuda~cxx~cxx_exceptions+gpfs~internal-hwloc~java~legacylaunchers~lustre~memchecker~pmi~pmix+romio~singularity~sqlite3+static~thread_multiple+vt+wrapper-rpath fabrics=none schedulers=none arch=darwin-monterey-skylake
[+]          ^hwloc@2.7.0%apple-clang@13.0.0~cairo~cuda~gl~libudev+libxml2~netloc~nvml~opencl~pci~rocm+shared arch=darwin-monterey-skylake
[+]              ^libxml2@2.9.12%apple-clang@13.0.0~python arch=darwin-monterey-skylake
[+]                  ^libiconv@1.16%apple-clang@13.0.0 libs=shared,static arch=darwin-monterey-skylake
[+]                  ^pkgconf@1.8.0%apple-clang@13.0.0 arch=darwin-monterey-skylake
[+]                  ^xz@5.2.5%apple-clang@13.0.0~pic libs=shared,static arch=darwin-monterey-skylake
[+]                  ^zlib@1.2.11%apple-clang@13.0.0+optimize+pic+shared arch=darwin-monterey-skylake
[+]              ^ncurses@6.2%apple-clang@13.0.0~symlinks+termlib abi=none patches=3970661 arch=darwin-monterey-skylake
[+]          ^libevent@2.1.12%apple-clang@13.0.0+openssl arch=darwin-monterey-skylake
[+]              ^openssl@1.1.1m%apple-clang@13.0.0~docs certs=system arch=darwin-monterey-skylake
[+]                  ^perl@5.34.0%apple-clang@13.0.0+cpanm+shared+threads arch=darwin-monterey-skylake
[+]                      ^berkeley-db@18.1.40%apple-clang@13.0.0+cxx~docs+stl patches=9f5b54a,b231fcc arch=darwin-monterey-skylake
[+]                      ^bzip2@1.0.8%apple-clang@13.0.0~debug~pic+shared arch=darwin-monterey-skylake
[+]                          ^diffutils@3.8%apple-clang@13.0.0 arch=darwin-monterey-skylake
[+]                      ^gdbm@1.19%apple-clang@13.0.0 arch=darwin-monterey-skylake
[+]                          ^readline@8.1%apple-clang@13.0.0 arch=darwin-monterey-skylake
[+]          ^openssh@8.8p1%apple-clang@13.0.0 arch=darwin-monterey-skylake
[+]              ^libedit@3.1-20210216%apple-clang@13.0.0 arch=darwin-monterey-skylake
```


### Error message

<details><summary>Error message</summary><pre>
6 errors found in build log:
     6474    mv -f .deps/rdft2-rank-geq2-transposed.Tpo .deps/rdft2-rank-geq2-transposed.Plo
     6475    mv -f .deps/rdft2-solve.Tpo .deps/rdft2-solve.Plo
     6476    mv -f .deps/rdft2-problem.Tpo .deps/rdft2-problem.Plo
     6477    mv -f .deps/api.Tpo .deps/api.Plo
     6478    /bin/sh ../libtool  --tag=CC   --mode=link /Users/eschnett/src/spack/opt/spack/darwin-monterey-skylake/gcc-11.2.0/open
             mpi-4.1.2-7poxc72sktaq4bllursf2spuanjv7sw2/bin/mpicc  -O3 -fomit-frame-pointer -mtune=native -malign-double -fstrict-a
             liasing -fno-schedule-insns -Wa,-q -Wl,-no_compact_unwind -version-info 9:10:6  -o libfftw3_mpi.la -rpath /Users/eschn
             ett/src/spack/opt/spack/darwin-monterey-skylake/gcc-11.2.0/fftw-3.3.10-beequmaebpecoqrwnyoyvr32v66ymrv2/lib any-true.l
             o api.lo block.lo choose-radix.lo conf.lo dtensor.lo rearrange.lo wisdom-api.lo f03-wrap.lo transpose-alltoall.lo tran
             spose-pairwise.lo transpose-recurse.lo transpose-problem.lo transpose-solve.lo dft-serial.lo dft-rank-geq2.lo dft-rank
             -geq2-transposed.lo dft-rank1.lo dft-rank1-bigvec.lo dft-problem.lo dft-solve.lo rdft-serial.lo rdft-rank-geq2.lo rdft
             -rank-geq2-transposed.lo rdft-rank1-bigvec.lo rdft-problem.lo rdft-solve.lo rdft2-serial.lo rdft2-rank-geq2.lo rdft2-r
             ank-geq2-transposed.lo rdft2-problem.lo rdft2-solve.lo ../libfftw3.la   -lm
     6479    libtool: link: /Users/eschnett/src/spack/lib/spack/env/gcc/gcc -dynamiclib  -o .libs/libfftw3_mpi.3.dylib  .libs/any-t
             rue.o .libs/api.o .libs/block.o .libs/choose-radix.o .libs/conf.o .libs/dtensor.o .libs/rearrange.o .libs/wisdom-api.o
              .libs/f03-wrap.o .libs/transpose-alltoall.o .libs/transpose-pairwise.o .libs/transpose-recurse.o .libs/transpose-prob
             lem.o .libs/transpose-solve.o .libs/dft-serial.o .libs/dft-rank-geq2.o .libs/dft-rank-geq2-transposed.o .libs/dft-rank
             1.o .libs/dft-rank1-bigvec.o .libs/dft-problem.o .libs/dft-solve.o .libs/rdft-serial.o .libs/rdft-rank-geq2.o .libs/rd
             ft-rank-geq2-transposed.o .libs/rdft-rank1-bigvec.o .libs/rdft-problem.o .libs/rdft-solve.o .libs/rdft2-serial.o .libs
             /rdft2-rank-geq2.o .libs/rdft2-rank-geq2-transposed.o .libs/rdft2-problem.o .libs/rdft2-solve.o   ../.libs/libfftw3.dy
             lib -lm  -O3 -mtune=native -malign-double -Wl,-no_compact_unwind   -install_name  /Users/eschnett/src/spack/opt/spack/
             darwin-monterey-skylake/gcc-11.2.0/fftw-3.3.10-beequmaebpecoqrwnyoyvr32v66ymrv2/lib/libfftw3_mpi.3.dylib -compatibilit
             y_version 10 -current_version 10.10 -Wl,-single_module
  >> 6480    Undefined symbols for architecture x86_64:
     6481      "_MPI_Abort", referenced from:
     6482          _fftw_mpi_gather_wisdom in wisdom-api.o
     6483          _fftw_mpi_broadcast_wisdom in wisdom-api.o
     6484      "_MPI_Allreduce", referenced from:
     6485          _fftw_mpi_any_true in any-true.o
     6486          _wisdom_ok_hook in api.o

     ...

     6572      "_ompi_mpi_unsigned", referenced from:
     6573          _wisdom_ok_hook in api.o
     6574      "_ompi_mpi_unsigned_long", referenced from:
     6575          _fftw_mpi_gather_wisdom in wisdom-api.o
     6576          _fftw_mpi_broadcast_wisdom in wisdom-api.o
     6577    ld: symbol(s) not found for architecture x86_64
  >> 6578    collect2: error: ld returned 1 exit status
  >> 6579    make[3]: *** [libfftw3_mpi.la] Error 1
  >> 6580    make[2]: *** [all] Error 2
  >> 6581    make[1]: *** [all-recursive] Error 1
  >> 6582    make: *** [all] Error 2

</pre></details>

I believe the problem is related to `libtool`. `libtool` is configured without MPI, and saves as `CC` the path to the actual compiler. Later, when building MPI code, fftw redefines `CC` to `$MPICC`. `libtool` seems to ignore this setting, and then tires to create a dynamic library while calling `$CC`, which fails since the `-lmpi` option is missing.

The same seems to happen on both Linux and macOS. However, the build finishes fine on Linux. Maybe the dynamic linker behaves different on Linux, and is not flagging missing external symbols when creating a dynamic library?

This problem was introduced fairly recently; I used to be able to build `fftw` without problems.

### Information on your system

$ spack debug report
* **Spack:** 0.17.1-1357-d62b8f0bf3
* **Python:** 3.8.9
* **Platform:** darwin-monterey-skylake
* **Concretizer:** clingo


### Additional information

[spack-build-out.txt](https://github.com/spack/spack/files/8143245/spack-build-out.txt)
[spack-build-env.txt](https://github.com/spack/spack/files/8143247/spack-build-env.txt)



### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/fftw/package.py
+++ b/var/spack/repos/builtin/packages/fftw/package.py
@@ -100,6 +100,7 @@ class FftwBase(AutotoolsPackage):
         # MPI. libtool then calls wrong linker to create shared libraries
         # (it calls `$CC` instead of `$MPICC`), and MPI symbols
         # remain undefined because `-lmpi` is not passed to the linker.
+        # https://github.com/FFTW/fftw3/issues/274
         # https://github.com/spack/spack/issues/29224
         if self.spec.satisfies('+mpi') and self.spec.satisfies('platform=darwin'):
             env.append_flags('LIBS', self.spec['mpi'].libs.ld_flags)

--- a/var/spack/repos/builtin/packages/fftw/package.py
+++ b/var/spack/repos/builtin/packages/fftw/package.py
@@ -100,7 +100,8 @@ class FftwBase(AutotoolsPackage):
         # MPI. libtool then calls wrong linker to create shared libraries
         # (it calls `$CC` instead of `$MPICC`), and MPI symbols
         # remain undefined because `-lmpi` is not passed to the linker.
-        if self.spec.satisfies('+mpi'):
+        # https://github.com/spack/spack/issues/29224
+        if self.spec.satisfies('+mpi') and self.spec.satisfies('platform=darwin'):
             env.append_flags('LIBS', self.spec['mpi'].libs.ld_flags)
 
     def configure(self, spec, prefix):

--- a/var/spack/repos/builtin/packages/fftw/package.py
+++ b/var/spack/repos/builtin/packages/fftw/package.py
@@ -101,7 +101,7 @@ class FftwBase(AutotoolsPackage):
         # (it calls `$CC` instead of `$MPICC`), and MPI symbols
         # remain undefined because `-lmpi` is not passed to the linker.
         if self.spec.satisfies('+mpi'):
-            env.append_flags('LIBS', '-lmpi')
+            env.append_flags('LIBS', self.spec['mpi'].libs.ld_flags)
 
     def configure(self, spec, prefix):
         # Base options

--- a/var/spack/repos/builtin/packages/fftw/package.py
+++ b/var/spack/repos/builtin/packages/fftw/package.py
@@ -96,6 +96,12 @@ class FftwBase(AutotoolsPackage):
                 'CXXFLAGS', self.spec['llvm-openmp'].headers.include_flags)
             env.append_flags(
                 'LDFLAGS', self.spec['llvm-openmp'].libs.ld_flags)
+        # FFTW first configures libtool without MPI, and later uses it with
+        # MPI. libtool then calls wrong linker to create shared libraries
+        # (it calls `$CC` instead of `$MPICC`), and MPI symbols
+        # remain undefined because `-lmpi` is not passed to the linker.
+        if self.spec.satisfies('+mpi'):
+            env.append_flags('LIBS', '-lmpi')
 
     def configure(self, spec, prefix):
         # Base options

